AppCenter/AppCenter/Internals/Storage/MSDBStorage.m

```diff
@@ -231,8 +211,75 @@ - (void)deleteDatabase {
   }
 }
 
-- (void)migrateDatabase:(void *)__unused db
-            fromVersion:(NSUInteger)__unused version {
+- (void)migrateDatabase:(void *)__unused db fromVersion:(NSUInteger)__unused version {
+}
+
+- (void)setMaxStorageSize:(long)sizeInBytes completionHandler:(nullable void (^)(BOOL))completionHandler {
+
+  // Check the current number of pages in the database to determine whether the requested size will shrink the database.
+  NSArray<NSArray *> *rows = [self executeSelectionQuery:@"PRAGMA page_count;"];
+  int currentPageCount = [rows[0][0] intValue];
+  MSLogDebug([MSAppCenter logTag], @"Found %i pages in the database.", currentPageCount);
+  int requestedMaxPageCount = [MSDBStorage numberOfPagesInBytes:sizeInBytes];
+  if (currentPageCount > requestedMaxPageCount) {
+    MSLogWarning([MSAppCenter logTag],
+                 @"Cannot change database size to %ld bytes as it would cause a loss of data. The"
+                 " default maximum size of %ld will not be overridden.",
+                 sizeInBytes,
+                 kMSDefaultDatabaseSizeInBytes);
+    if (completionHandler) {
+      completionHandler(NO);
+    }
+    return;
+  }
+  self.maxPageCount = requestedMaxPageCount;
+
+  // Capture the completion handler in the property. It will be executed once when opening the DB connection.
+  self.maxStorageSizeCompletionHandler = completionHandler;
+}
+
+- (sqlite3 *)openDatabaseAtFileURL:(NSURL *)fileURL withMaxPageCount:(int)maxPageCount withResult:(int *)result {
+  sqlite3 *db = NULL;
+  *result = sqlite3_open_v2([[fileURL absoluteString] UTF8String],
+                            &db,
+                            SQLITE_OPEN_READWRITE | SQLITE_OPEN_CREATE | SQLITE_OPEN_URI,
+                            NULL);
+  if (*result != SQLITE_OK) {
+    MSLogError([MSAppCenter logTag], @"Failed to open database with result: %d.", *result);
+    return NULL;
+  }
+  MSLogVerbose([MSAppCenter logTag], @"Opened database.");
+  NSString *statement = [NSString stringWithFormat:@"PRAGMA max_page_count = %i;", maxPageCount];
+  char *errorMessage;
+  *result = sqlite3_exec(db, [statement UTF8String], NULL, NULL, &errorMessage);
+  if (*result != SQLITE_OK) {
+    errorMessage = errorMessage ? errorMessage : "(nil)";
+    NSString *printableErrorMessage = [NSString stringWithCString:errorMessage encoding:NSUTF8StringEncoding];
+    MSLogWarning([MSAppCenter logTag], @"Failed to open database with specified"
+                                       "maximum size constraint. Error message:"
+                                       " %@", printableErrorMessage);
+
+    // Call the callback for setting the maximum storage size only once.
+    dispatch_once(&setMaxStorageSizeOnceToken, ^{
+      if (self.maxStorageSizeCompletionHandler) {
+        self.maxStorageSizeCompletionHandler(NO);
+      }
+    });
+  } else {
+    MSLogVerbose([MSAppCenter logTag], @"Database has maximum page count of %i.", maxPageCount);
+
```

Comment:
In Android we log a warning message like "Could change database size but is slightly larger than expected (next multiple of 4KB)..." when we set the max storage size to the next multiple of 4KB (see https://github.com/Microsoft/AppCenter-SDK-Android/blob/develop/sdk/appcenter/src/main/java/com/microsoft/appcenter/utils/storage/DatabaseManager.java#L530-L532).

For consistency purpose, we should do the same in iOS if we're not already doing that.

Reply:
Good idea.

Reply:
So I made a change to this https://github.com/Microsoft/AppCenter-SDK-Apple/pull/1120/commits/8abe706a4351d465d0935246e567e7ea57ebbcf2

I'm not sure if we want to keep it as it will be called a lot. Note that iOS is not reusing one connection but opening/closing them all the time. This means the max size will be set time and again, so the warning will be logged a lot.
I am thinking of changing it to verbose, what do you think?

Reply:
@ElektrojungeMS, okay, either `MSLogInfo` or `MSLogVerbose` should be fine.